### ISSUE TITLE
"󰈙" character is an example of one that my tmux built from source gets garbled on

### ISSUE THREAD

**Issue by unphased** [tmux/tmux#3799]
### Issue description

The character in question is an example of a nerd font codepoint. This particular one is used in neovim to indicate a file with `.txt` file type, it is rendered in a nerd font font with a file icon. 

Whenever I open neovim inside tmux, it tends to render as some sort of a zero width character, and I notice it also messes with the readline implementation in zsh and bash in various ways whenever the character is entered in any way on the command line.

### Required information

* tmux version (`tmux -V`) = `tmux next-3.4`.
* Platform (`uname -sp`) = `Darwin arm`.
* $TERM inside and outside of tmux (`echo $TERM`) = outside `alacritty` inside `screen-256color` (note i set zero config).
* Logs from tmux (`tmux kill-server; tmux -vv new`). Please find attached:

[tmux-out-5099.log](https://github.com/tmux/tmux/files/13852591/tmux-out-5099.log)
[tmux-server-5099.log](https://github.com/tmux/tmux/files/13852593/tmux-server-5099.log)
[tmux-client-5097.log](https://github.com/tmux/tmux/files/13852594/tmux-client-5097.log)

The logs correspond to this following minimal repro run, which involves echoing this character in bash using two different approaches and sleeping for 2 sec before exiting bash so you can visually confirm the character has been written out twice:

`tmux -f /dev/null -vv -L test new-session bash -c "echo $'\xf3\xb0\x88\x99'; echo '󰈙'; sleep 2;" `

I am not sure if this is sufficient to produce enough information in the server log so let me know if I should test with a different method.

**Comment by krader1961** [comment]
FYI, mechanisms like [Nerd Fonts](https://www.nerdfonts.com/) rely on abusing the Unicode private use area code points. Those private use code points do not have a well defined length. This feedback might be irrelevant to your specific problem. I mention it because it is a typical reason for the behavior you are complaining about.

**Comment by unphased** [comment]
Yeah I do understand that. It's just that I hope there's some tweak I can do to compile a tmux that will fall back to single width instead of zero width for unknown characters like this. 

I have to assume there is a good group of folks out there like us who use some fancy neovim plugins inside tmux. On remote servers. This is how I write all of my code, sure I do it mostly locally and not on a remote server but being able to use the same system (relying on tmux as persistence layer) everywhere is powerful.

**Comment by krader1961** [comment]
> I have to assume there is a good group of folks out there like us who use some fancy neovim plugins inside tmux.

It seems to me you are relying on a mechanism that is abusing Unicode PUA codepoints. Which is inherently impossible to handle correctly without providing hints to each program that interprets those PUA codepoints. Before proceding further I think there needs some clarification regarding the problematic code point. That information might be in the provided logs but it would be nice if the O.P. told us the value of the codepoint rather than forcing us to extract that datum from the log files.

**Comment by nicm** [comment]
I worked before on a way to override the system width for individual or ranges of codepoints but it turned out not to be needed, something like that would be a good idea.

IIRC tmux assumes width one for these not zero.

**Comment by unphased** [comment]
One thing I'll note is that I do not experience this issue on Linux. the plot thickens slightly.

**Comment by unphased** [comment]
I augmented my test case with some identifiable strings. 

1704689806.632080 screen_write_start_pane: size 118x48, pane %0 (at 0,0)
1704689806.632083 input_parse_buffer: %0 ground, 34 bytes: abcdef_ghijkl\363\260\210\231\r\nabcabcabc\363\260\210\231\r\n
1704689806.632087 screen_write_collect_end: 13 abcdef_ghijkl (at 0,0)
1704689806.632090 UTF-8 \363\260\210\231 is 000F0219
1704689806.632093 wcwidth(000F0219) returned 2
1704689806.632095 input_top_bit_set 4 '\363\260\210\231' (width 2)
1704689806.632098 /dev/ttys055: abcdef_ghijkl
1704689806.632103 screen_write_collect_flush: flushed 1 items (screen_write_collect_add)
1704689806.632106 screen_write_cell: new padding at 14,0
1704689806.632109 utf8_from_data: (0 0 ) -> 20000021
1704689806.632111 utf8_put_item: added \363\260\210\231 = 0

Looks like the problem is a derivation of width 2 for the char in question. @nicm got any tips?

**Comment by unphased** [comment]
Doing a small amount of research, it may be better to use something like described in this comment https://github.com/alacritty/alacritty/issues/265#issuecomment-272009517 which is going to produce correct unicode width determination for modern usage. 

It does seem like in order to prevent failures, compatible width calculation ought to be used all the way down the chain (nvim, tmux, terminal emulator)...

**Comment by unphased** [comment]
One more thing that is throwing me for a loop... 

If I ssh to my linux box from INSIDE tmux on macos the issues with these characters go away. they render properly and don't screw up terminal layout. That is mildly perplexing. maybe running tmux inside tmux causes the outer tmux not to reapply character width math somehow.

**Comment by unphased** [comment]
I have resolved this by rebuilding tmux on my mac after installing utf8proc with brew and then specifying `./configure --enable-utf8proc`.
